storage/client.go

```diff
@@ -75,17 +75,27 @@ type DefaultSender struct {
 
 // Send is the default retry strategy in the client
 func (ds *DefaultSender) Send(c *Client, req *http.Request) (resp *http.Response, err error) {
-	b := []byte{}
+	var reader *bytes.Reader
 	if req.Body != nil {
-		b, err = ioutil.ReadAll(req.Body)
+		// copy the request body to a local buffer; this is to
+		// support retry logic as the underlying transport will
+		// always close the request body (success or failure).
+		b := make([]byte, req.ContentLength)
+		_, err = io.ReadFull(req.Body, b)
 		if err != nil {
 			return resp, err
 		}
+		reader = bytes.NewReader(b)
 	}
 
 	for attempts := 0; attempts < ds.RetryAttempts; attempts++ {
-		if len(b) > 0 {
-			req.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+		if reader != nil {
+			// move the reader back to the beginning of the buffer
+			_, err = reader.Seek(0, io.SeekStart)
+			if err != nil {
+				return resp, err
+			}
+			req.Body = ioutil.NopCloser(reader)
```

Comment:
I _believe_ you can pull this line outside of the loop, which will result in N fewer allocations of a blank NopCloser.